//---------------------------------------------------------------------
// Copyright 2020 National Instruments
//
// Licensed under the MIT License;
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://mit-license.org/
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//---------------------------------------------------------------------

//---------------------------------------------------------------------
//---------------------------------------------------------------------
syntax = "proto3";

//---------------------------------------------------------------------
//---------------------------------------------------------------------
option java_multiple_files = true;
option java_package = "labview.measurementservice";
option java_outer_classname = "MeasurementService";
option objc_class_prefix = "LVMS";

//---------------------------------------------------------------------
//---------------------------------------------------------------------
package measurementservice;

//---------------------------------------------------------------------
// The MeasurementService service definition.
//---------------------------------------------------------------------
service MeasurementService {
  
  rpc Invoke (InvokeRequest) returns (InvokeResponse) {}
  
  rpc Query (QueryRequest) returns (QueryResponse) {}
  
  rpc Register (RegistrationRequest) returns (stream ServerEvent) {}
  
  rpc OpenPWMSession (PWMOpenSessionRequest) returns (PWMOpenSessionResponse) {}
  
  rpc StartPWM (PWMStartRequest) returns (PWMStartResponse) {}
  
  rpc ModifyPWM (PWMModifyRequest) returns (PWMModifyResponse) {}
  
  rpc StopPWM (PWMStopRequest) returns (PWMStopResponse) {}
  
  rpc ClosePWMSession (PWMCloseSessionRequest) returns (PWMCloseSessionResponse) {}
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message ErrorOut {
  int32 errCode = 1;
  string errMessage = 2;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message PWMOpenSessionRequest {
  string PWMlines = 1;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message PWMOpenSessionResponse {
  ErrorOut error = 1;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message PWMStartRequest {
  bool PWMInitialState = 1;
  bool PWMInvertLines = 2;
  int32 PWMFrequency = 3;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message PWMStartResponse {
  bool PWMInitialState = 1;
  bool PWMInvertLines = 2;
  int32 PWMFrequency = 3;
  ErrorOut error = 4;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message PWMModifyRequest {
  int32 PWMFrequency = 1;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message PWMModifyResponse {
  int32 PWMFrequency = 1;
  ErrorOut error = 2;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message PWMStopRequest {
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message PWMStopResponse {
  ErrorOut error = 1;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message PWMCloseSessionRequest {
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message PWMCloseSessionResponse {
  ErrorOut error = 1;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message InvokeRequest {
  string command = 1;
  string parameter = 2;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message InvokeResponse {
  int32 status = 1;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message QueryRequest {
  string query = 1;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message QueryResponse {
  string message = 1;
  int32 status = 2;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message RegistrationRequest {
  string eventName = 1;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message ServerEvent {
  string eventData = 1;
  int32 serverId = 2;
  int32 status = 3;
}
